runtime/src/main/cpp/Exceptions.cpp

```diff
@@ -206,10 +213,24 @@ void ThrowException(KRef exception) {
 #endif
 }
 
-void ReportUnhandledException(KRef e);
+OBJ_GETTER(Kotlin_setUnhandledExceptionHook, KRef hook) {
+  RETURN_RESULT_OF(SwapRefLocked,
+    &currentUnhandledExceptionHook, currentUnhandledExceptionHook, hook, &currentUnhandledExceptionHookLock);
+}
+
+void OnUnhandledException(KRef throwable) {
+  ObjHolder handlerHolder;
+  auto* handler = SwapRefLocked(&currentUnhandledExceptionHook, currentUnhandledExceptionHook, nullptr,
+     &currentUnhandledExceptionHookLock, handlerHolder.slot());
+  if (handler == nullptr) {
+    ReportUnhandledException(throwable);
+  } else {
+    ExceptionReporterLaunchpad(handler, throwable);
+  }
+}
 
-RUNTIME_NORETURN void TerminateWithUnhandledException(KRef e) {
-  ReportUnhandledException(e);
+RUNTIME_NORETURN void TerminateWithUnhandledException(KRef throwable) {
+  OnUnhandledException(throwable);
```

Comment:
It makes a lot of sense to prevent an exception to be thrown out of `OnUnhandledException`.